File: AppSets-Android/appsets/src/main/kotlin/xcj/app/appsets/im/BrokerTest.kt
package xcj.app.appsets.im

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import com.google.gson.Gson
import com.google.gson.JsonObject
import xcj.app.appsets.account.LocalAccountManager
import xcj.app.appsets.im.message.IMMessage
import xcj.app.appsets.settings.ModuleConfig
import xcj.app.starter.android.util.PurpleLogger
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

object BrokerTest {

    private const val TAG = "BrokerTest"

    private val broker = RabbitMQBroker()

    val imOnLineState: MutableState<IMOnlineState> = mutableStateOf(IMOnlineState.Offline)

    @OptIn(ExperimentalEncodingApi::class)
    suspend fun start() {
        PurpleLogger.current.d(TAG, "start")
        if (!LocalAccountManager.isLogged()) {
            PurpleLogger.current.d(
                TAG,
                "start, failed! because of use not login! return"
            )
            return
        }

        val appConfig = ModuleConfig.moduleConfiguration
        if (appConfig.imBrokerProperties.isEmpty()) {
            PurpleLogger.current.d(
                TAG,
                "start, failed!" +
                        " because of imBrokerProperties isNullOrEmpty, return"
            )
            return
        }
        PurpleLogger.current.d(
            TAG,
            "start, AppConfig.appConfiguration.imBrokerProperties:${appConfig.imBrokerProperties}"
        )

        val decodeConfig =
            Base64.decode(appConfig.imBrokerProperties.toByteArray())
                .decodeToString()
        PurpleLogger.current.d(
            TAG,
            "start, decodeConfig:$decodeConfig"
        )
        if (decodeConfig.isEmpty()) {
            PurpleLogger.current.d(
                TAG,
                "start, failed!" +
                        " decode RabbitProperties get null, return"
            )
            return
        }
        runCatching {
            val jsonObject = Gson().fromJson<JsonObject>(
                decodeConfig,
                JsonObject::class.java
            )
            PurpleLogger.current.d(
                TAG,
                "start, configMap:${jsonObject}"
            )
            val rabbitMqBrokerConfig = RabbitMQBrokerConfig(
                RabbitMQBrokerProperty(
                    host = jsonObject["rabbit-host"].asString,
                    port = jsonObject["rabbit-port"].asInt,
                    username = jsonObject["rabbit-admin-username"].asString,
                    password = jsonObject["rabbit-admin-password"].asString,
                    virtualHost = jsonObject["rabbit-virtual-host"].asString,
                    queuePrefix = jsonObject["queue-prefix"].asString,
                    routingKeyPrefix = jsonObject["routing-key-prefix"].asString,
                    groupExchangePrefix = "one2many-fanout-",
                    groupExchangeParent = "one2many-topic",
                    groupRootExchange = "one2many-fanout-root",
                    groupSubRootExchange = "one2many-fanout-subroot"
                )
            )
            rabbitMqBrokerConfig
        }.onSuccess { config ->
            broker.bootstrap(config)
        }.onFailure {
            PurpleLogger.current.d(
                TAG,
                "start, failed!" +
                        " RabbitProperties map get failed, ${it.message}\n ${it.stackTraceToString()}"
            )
        }
    }

    suspend fun sendMessage(imObj: IMObj, imMessage: IMMessage) {
        PurpleLogger.current.d(TAG, "sendMessage")
        broker.sendMessage(imObj, imMessage)
    }

    suspend fun close() {
        broker.close()
    }

    suspend fun updateImGroupBindIfNeeded() {
        broker.updateImGroupBindIfNeeded()
    }

}

File: AppSets-iOS/AppSets/com/app/appsets/im/BrokerTest.swift
//
//  BrokerTest.swift
//  AppSets
//
//  Created by caiju Xu on 2024/7/9.
//

import Foundation

@Observable
class BrokerTest {
    
    private static let TAG = "BrokerTest"
    
    public static let Instance = BrokerTest()
    
    private let broker: RabbitMQBroker = RabbitMQBroker()
    
    var isOnline: Bool = false
    
    func changeOnline(_ new:Bool){
        DispatchQueue.main.async {
            self.isOnline = new
        }
    }
    
    private init(){
        
    }
    
    func start() {
        PurpleLogger.current.d(BrokerTest.TAG, "start")
        Task{
            try await Task.sleep(nanoseconds: 2_000_000_000)
            if !LocalAccountManager.Instance.isLogged() {
                PurpleLogger.current.d(BrokerTest.TAG, "start, failed! because of use not login! return")
                return
            }
            
            let appConfig = AppConfig.Instance.appConfiguration
            if (String.isNullOrEmpty(appConfig.imBrokerProperties)){
                PurpleLogger.current.d(BrokerTest.TAG, "start, failed! because of imBrokerProperties isNullOrEmpty, return")
                return
            }
            PurpleLogger.current.d(BrokerTest.TAG, "start, parse config:\(appConfig.imBrokerProperties)")
            let decodedConfig = String.unWrappeBase64(appConfig.imBrokerProperties)
            PurpleLogger.current.d(BrokerTest.TAG, "start, decodedConfig:\(decodedConfig)")
            guard let configMap = String.decodeJSONFromString(jsonString: decodedConfig!) as? [String:Any] else{
                return
            }
        
            let userInfo = LocalAccountManager.Instance.userInfo
            let property = RabbitMqBrokerProperty(
                uid: userInfo.uid,
                userExchangeGroups: "G7775",
                host: configMap["rabbit-host"] as? String ?? "",
                port: configMap["rabbit-port"] as? Int ?? 5672,
                username: configMap["rabbit-admin-username"] as? String ?? "",
                password: configMap["rabbit-admin-password"] as? String ?? "",
                virtualHost: configMap["rabbit-virtual-host"] as? String ?? "",
                queuePrefix: configMap["queue-prefix"] as? String ?? "",
                routingKeyPrefix: configMap["routing-key-prefix"] as? String ?? "",
                groupExchangePrefix: "one2many-fanout-",
                groupExchangeParent: "one2many-topic",
                groupRootExchange: "one2many-fanout-root",
                groupSubRootExchange: "one2many-fanout-subroot"
            )
            let config = RabbitMQBrokerConfig(property)
            
            broker.bootstrap(config)
        }
       
    }
    
    func sendMessage(_ imObj:ImObj, _ imMessage: any ImMessage) -> Bool {
        PurpleLogger.current.d(BrokerTest.TAG, "sendMessage")
        return broker.sendMessage(imObj, imMessage)
    }
    
    func close() {
        broker.close()
    }
    
    func updateImGroupBindIfNeeded() {
        broker.updateImGroupBindIfNeeded()
    }
}
